DM-15775: Move parts of l1dbproto to new package

andy-slac opened the pull request:
This is completely new package which contains few of the modules that originally appeared in `lsst-dm/l1dbproto`. I think I managed to preserve more or less complete commit history for those modules. The module have been moved here with original names (l1db*) and then renamed to `PPDB`, including the classes in the modules. Clients will need some effort to rename modules/classes on their side too.

All unit tests run OK, I have also tested this with my `ap_proto` application (which is still in `l1dbproto`).

andy-slac commented on COPYRIGHT:
```diff
@@ -0,0 +1 @@
+Copyright 2017-2018 The Board of Trustees of the Leland Stanford Junior University, through SLAC National Accelerator Laboratory
```
Sure, sorry, I forgot about those edits.

andy-slac commented on cfg/l1db-mysql.py:
```diff
@@ -1,57 +0,0 @@
-import lsst.l1dbproto.l1db
-assert type(config)==lsst.l1dbproto.l1db.L1dbConfig, 'config is of type %s.%s instead of lsst.l1dbproto.l1db.L1dbConfig' % (type(config).__module__, type(config).__name__)
```
I agree, but this is beyond our control. Github decides for itself which changes to treat as simple renaming vs. delete+add. I guess this depends on how many lines were changed/added. I could try to play some games splitting one commit in multiple but I don't want to spend time on that now.

timj replied:
I have no idea what the commit history is, but git will only see a rename if the code is removed and added in the same commit (git mv does exactly this but is not required for git to see the move). Splitting the commit into multiple bits pretty much guarantees that git won't see a rename.

andy-slac replied:
Well I guess my mistake was renaming it (with `git mv`) and then editing its contents before committing which probably confuses git(hub). Doing it in two commits, one for rename, and one for editing should unconfuse things.

andy-slac replied:
I have split that commit, I think it should look better when I force-push.

andy-slac commented on python/lsst/dax/ppdb/ppdb.py:
```diff
@@ -129,6 +134,13 @@ class L1dbConfig(pexConfig.Config):
+                            default=True)
+    sql_echo = Field(dtype=bool,
+                     doc="If True then pass SQLAlchemy echo option.",
+                     default=False)
```
This was added when I worked on unit tests, in-memory sqlite gave me some trouble (due to connection management) so I needed to debug it (hence `echo` option) and to solve the problem I also needed to enable connection pool. Connection pool has to be disabled for my `ap_proto` script which depends on forking so I have to have this new option to control connection pooling. Echo option could be useful in the future so I decided to keep it too.

andy-slac commented on python/lsst/dax/ppdb/ppdb.py:
```diff
@@ -269,6 +279,9 @@ def lastVisit(self):
 
+        This method is only used by ap_proto and is not intended for
+        production pipelines.
+
```
`ap_proto` is a script in `l1dbproto` which simulates AP pipeline (I use it for my studies). I presume production code will have some different mechanism for tracking visit numbers, so this method is not needed for production at all. Maybe I should move this method back to `l1dbproto`.

kfindeisen replied:
Maybe? If you do keep it, can you make it clear that `ap_proto` is from `l1dbproto`?

andy-slac commented on python/lsst/dax/ppdb/ppdb.py:
```diff
@@ -392,15 +408,18 @@ def getDiaSourcesInRegion(self, pixel_ranges, dt):
         -------
-        `afw.table.BaseCatalog` instance
+        `afw.table.SourceCatalog` instance, `None` if ``read_sources_months``
+        configuration parameter is set to 0.
```
OK, naming things that do not need names is my favorite pastime 😄

andy-slac commented on python/lsst/dax/ppdb/ppdb.py:
```diff
@@ -392,15 +408,18 @@ def getDiaSourcesInRegion(self, pixel_ranges, dt):
         if self.config.read_sources_months == 0:
             _LOG.info("Skip DiaSources fetching")
-            return []
+            return None
```
This change is related to me finally adding some unit tests and making sure that methods return reasonable values. It should have been done earlier I guess to avvoid potential issues on ap_association side, but now it's not too late either.

This method normally returns a catalog, in this case I don't want to return an empty catalog because I don't know how to build correct schema for it (though it can be built from a table schema I want to avoid querying the table just to build afw schema). And I think returning `None` is better that returning `[]` because catalog is definitely not a list, so if you need to check for some special return value it is better to check for `None` than for `[]`. Anyways, method docstring now says you should expect either catalog or `None`.

kfindeisen replied:
I still don't like it, but I'll concede that a list would be equally bad.

andy-slac replied:
Well, I think `None` is useful in a sense that you can distinguish cases when PPDB was looking for some data but did not find any (in which case empty catalog is returned) vs case when PPDB says "you want some data but I was configured to not even look for it".

andy-slac commented on python/lsst/dax/ppdb/ppdbSchema.py:
```diff
@@ -108,7 +120,7 @@ def _add_suffixes(element, compiler, **kw):
 @compiles(CreateIndex, "oracle")
-def _add_suffixes(element, compiler, **kw):
+def _add_suffixes_idx(element, compiler, **kw):
     """Add all needed suffixed for Oracle CREATE INDEX statement"""
```
I'll add a link to the corresponding SQLALchemy docs, just describing parameters here will not make things obvious.

andy-slac commented on python/lsst/dax/ppdb/ppdbSchema.py:
```diff
@@ -140,35 +148,32 @@ class L1dbSchema(object):
         DiaForcedSource table instance
     visits : `sqlalchemy.Table`
-        L1DbProtoVisits table instance
+        PpdbProtoVisits table instance
```
Interesting. Looks like style guide is changing faster than I can type docstrings 😒 I just sphinxed that source and output looks perfectly fine to me. Anyways, who am I to contradict wise ⬛️ guys 😏, will have to fix that now.

andy-slac replied:
Hmm, if I switch to the new style then sphinx does not generate anything for those attributes. I'm going to leave things as they were, this can be sorted out later.

andy-slac commented on python/lsst/dax/ppdb/ppdbSchema.py:
```diff
@@ -200,10 +205,13 @@ def __init__(self, engine, dia_object_index, dia_object_nightly,
-            self._column_map = yaml.load(yaml_stream)
-            _LOG.debug("column map: %s", self._column_map)
+        if column_map:
+            with open(column_map) as yaml_stream:
```
I moved that message inside `if` and added different one in `else`.

andy-slac commented on python/lsst/dax/ppdb/timer.py:
```diff
@@ -1,31 +1,37 @@
+"""Module with methods to return timing information.
+
+This was developed as a part of prototype for performance studies. It
+could probably be removed in production system.
```
It is certainly useful now for my performance studies, which I'm still working on. And I think it may be OK to keep it for production as monitoring tool.

andy-slac commented on ups/dax_ppdb.table:
```diff
@@ -1,12 +1,11 @@
-setupRequired(db)
 setupRequired(pex_config)
 setupRequired(python)
-setupRequired(pyyaml)
```
Thanks, missed that somehow.

andy-slac commented on tests/test_ppdb.py:
```diff
@@ -0,0 +1,336 @@
+from lsst.dax.ppdb import (Ppdb, PpdbConfig, make_minimal_dia_object_schema,
+                           make_minimal_dia_source_schema)
+from lsst.sphgeom import Angle, Circle, HtmPixelization, Vector3d, UnitVector3d
+from lsst.geom import SpherePoint
```
I think I tried to understand how to convert Vector3d into ra/decl and the only way that I found is via SpherePoint. I probably need to spend some time to understand why there are few different geometry packages in LSST and try to learn the one true way to do things. LEt me leave this for later time after I finish this ticket.

kfindeisen replied:
"Why there are a few different packages" -- basically because different people wanted slightly different things at different times.

You may be interested in the [`sphgeom.LonLat`](http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/classlsst_1_1sphgeom_1_1_lon_lat.html) class.

andy-slac commented on tests/test_ppdb.py:
```diff
@@ -0,0 +1,336 @@
+    """make a catalog containing a bunch of DiaObjects inside pixel envelope.
+
+    The number of created records will be equal number of ranges (one object
+    per pixel range). Coodirnates of the created objects are not usable.
```
Not usable means don't look at them, their values are unspecified.

andy-slac commented on tests/test_ppdb.py:
```diff
@@ -0,0 +1,336 @@
+
+    @classmethod
+    def setUpClass(cls):
+        logging.basicConfig(level=logging.INFO)
```
Hmm, I guess I wanted to see some messages from unit test to understand what it does. I can probably disable it now. dax_ppdb uses Python logging everywhere, though this should not be an issue for the correctly-configured applications. Reminds me - we should probably revise logging levels of the messages at some point, what was good for my prototype may not be good for production.

timj replied:
Pure python packages are allowed (I'd almost say required but we aren't there yet) to use python logging. Tasks/PipelineTasks will configure lsst.log to accept python logging messages. I think the new pytest we are using allows logging level to be set for tests so I'll probably turn that on so that log messages always appear in tests.

andy-slac commented on tests/test_ppdb.py:
```diff
@@ -0,0 +1,336 @@
+        config = PpdbConfig(db_url="sqlite://",
+                            isolation_level="READ_UNCOMMITTED")
+        ppdb = Ppdb(config)
+        ppdb.makeSchema()
```
The only visible result in this case is that it does not throw any exceptions (I'll add this as a comment). The non-visible part of this is tested in `test_ppdbSchema.py`.

andy-slac commented on tests/test_ppdb.py:
```diff
@@ -0,0 +1,336 @@
+
+        # get sources by region
+        res = ppdb.getDiaSourcesInRegion(pixel_ranges, visit_time)
+        self.assertIs(res, None)
```
`getDiaObjects` always fetches something from database and knows how to build afw schema from that. For other methods there are cases when we cannot even get anything from the database so it becomes difficult of inefficient to build correct afw schema, and in that case I return `None`. If you are happy with me returning empty catalog with arbitrary schema instead that can be done too (just open new ticket for me).

andy-slac commented on tests/test_ppdb.py:
```diff
@@ -0,0 +1,336 @@
+        """Test for getting data from empty database.
+
+        All get() methods should return empty results, only useful for
+        checking that code is not broken.
```
While returning `None` for months=0 I know for sure that database contents does not matter. But if you will open that ticket for empty catalog instead of None, I'll need to add a new check for that case too.

andy-slac replied:
Well, I know current (expected) behavior and I write tests to keep that reproducible. If behavior changes or becomes more complex I'll sure need to update current tests and write more complex tests. It's a balance, it's hard to write tests for every possible imaginable case, I just try to test things that are important (during this migration).

andy-slac commented on tests/test_ppdb.py:
```diff
@@ -0,0 +1,336 @@
+        # sqlite does not support default READ_COMMITTED, for in-memory
+        # database have to use connection pool
+        config = PpdbConfig(db_url="sqlite://",
+                            isolation_level="READ_UNCOMMITTED")
```
I opted for READ_COMMITED as a default isolation level in PpdbConfig which seems reasonable for PPDB concurrency model (I think), but sqlite does not support that, it can only do READ_UNCOMMITED or SERIALIZABLE. I may need to rethink that and use "default" default instead of READ_COMMITED and configuse READ_COMMITED as optimization for production.

andy-slac commented on tests/test_ppdb.py:
```diff
@@ -0,0 +1,336 @@
+            record.set("coord_dec", obj['coord_dec'])
+            record.set("flags", 0)
+            record.set("pixelId", obj['pixelId'])
+            oids.append(obj['id'])
```
Again, if I need this in more than one place, I'll refactor this code too.

andy-slac commented on tests/test_ppdb.py:
```diff
@@ -0,0 +1,336 @@
+        schema = afwTable.Schema()
+        schema.addField("diaObjectId", "L")
+        schema.addField("ccdVisitId", "L")
+        schema.addField("flags", "L")
```
Are you suggesting a name for a new JIRA ticket? 😏 
Seriously if ap_association needs something like this we should add it.

kfindeisen replied:
I meant you've got a block of code that does something different from the rest of the method, and that's similar to other methods you've already written. But based on your previous answers I assume this one is also "once I write more tests".

andy-slac replied:
OK, I was assuming that you wanted to add new method to ppdbSchema module following two existing methods which are already there (make_minimal_dia_object_schema/make_minimal_dia_source_schema). If that is generally useful then we can add it there and I can use it here. Otherwise if there will be more than one use case for it in these tests I'll refactor it sooner or later into a local method.

andy-slac commented on tests/test_ppdbSchema.py:
```diff
@@ -0,0 +1,266 @@
+        self._assertTable(schema.sources, "DiaSource", 108)
+        self._assertTable(schema.forcedSources, "DiaForcedSource", 7)
+
+        # create shema using prefix
```
There is no real concern here about dropped tables. I prefer to run all these test sequentially just not to think about what happens if two of them could be run in parallel.

kfindeisen replied:
Well, I know from experience that SQLite in-memory databases are completely independent, even if you don't want them to be. 😠 

But seriously, with four test cases you're potentially running tests in parallel anyway, so splitting them up further won't cause any new problems.

andy-slac replied:
It is slightly more complicated with SQLAlchemy which keeps a connection pool so you can reuse the same database (in-memory sqlite indeed has separate database for each connection) by re-using connection from a pool. This is why I had to add connection_pool option to PpdbConfig to enable connection reuse which was disabled by default in my earlier implementation.

I guess I'm being too defensive about my decision to keep all in one method, maybe I will split it on a later tickets.

andy-slac commented on tests/test_ppdbSchema.py:
```diff
@@ -0,0 +1,266 @@
+        """Test for creating schemas.
+
+        Schema is defined in YAML files, some checks here depend on that
+        configuration and will need to be updated when configuration changes.
```
Indeed I will need to improve this. That was just a quick hack to assist me in migrating code and make sure that nothing breaks over this short period of time. For long term I think I should indeed switch to a stable set of shorter schema configs.

andy-slac commented on tests/test_ppdbSchema.py:
```diff
@@ -0,0 +1,266 @@
+        self.assertEqual(len(col_map), 7)
+        self.assertIsInstance(afw_schema, afwTable.Schema)
+        # afw table adds 4 columns compared to out standard schema
+        self.assertEqual(afw_schema.getFieldCount(), 7+4)
```
I'll fix that. And I have a comment for that just above 😃

andy-slac commented on tests/test_ppdbSchema.py:
```diff
@@ -0,0 +1,266 @@
+        afw_schema, col_map = schema.getAfwSchema("DiaSource")
+        self.assertEqual(len(col_map), 108)
+        self.assertIsInstance(afw_schema, afwTable.Schema)
+        self.assertEqual(afw_schema.getFieldCount(), 108)
```
I'm not sure that I want an extra overhead of creating completely new schema for each individual table. And it's more natural in this case to have a single schema (maybe just to make sure that a single schema contains info for all tables at once)

andy-slac commented:
I think I have addressed most of the comments (maybe not all of them adequately). I have just force-pushed all changes that I made in response to comments, unit tests still run OK, hopefully that does not break things on ap_association side (I'm going to check).

andy-slac commented:
`uv pip` is unhappy and I have no idea why. I'm going to ignore it for now.

timj commented:
I think the mypy problem is that `felis` should be called `lsst-felis` ?

timj commented on requirements.txt:
```diff
@@ -0,0 +1,7 @@
+git+https://github.com/lsst/dax_apdb@main#egg=lsst-dax-apdb
+git+https://github.com/lsst/utils@main#egg=lsst-utils
+git+https://github.com/lsst/resources@main#egg=lsst-resources
+git+https://github.com/lsst/felis@main#egg=lsst-felis
```
Can you modernize this syntax please?

```suggestion
lsst-felis @ git+https://github.com/lsst/felis@main
```
and the other lines. I see you are calling it `lsst-felis` and that does agree with the `pyproject.toml` in `felis` so that's a bit odd.

andy-slac replied:
I tried modern syntax and got this error, switched it back to old one, still the same error. Will switch back to modern.

timj commented on pyproject.toml:
```diff
@@ -0,0 +1,111 @@
+    "pandas",
+    "pyyaml >= 5.1",
+    "sqlalchemy",
+    "felis",
```
Maybe this is your problem.
```suggestion
    "lsst-felis",
```

andy-slac replied:
Ah, thanks! 🤦‍♂️
